Unix/server/server.c

```diff
@@ -166,6 +168,122 @@ static char** _DuplicateArgv(int argc, const char* argv[])
     return newArgv;
 }
 
+
+//
+// Duplicates and sets needed variables and removes the rest
+//
+static char** _DuplicateEnvp(const char* envp[])
+{
+    int i;
+
+    // Special env variables:
+    const char Krb5_Trace[]  = "KRB5_TRACE=/dev/stderr";
+    const char Krb5_KTName[] = "KRB5_KTNAME=";
+    const char Krb5_CCName[] = "KRB5_CCNAME=";
+    const char Ntlm_User_File[] = "NTLM_USER_FILE=";
+
+    int env_count = 0;
+    char *envitem = NULL;
+    for (envitem = (char*)(envp[0]); envitem; envitem = (char*)(envp[env_count]))
+    { 
+        env_count++;
+    }
+
+    char **newEnv = (char**)PAL_Malloc((env_count+3)*sizeof(char*));
+    if (!newEnv)
+    {
+        return NULL;
+    }
+
+    int new_env_count = 0;
+
+    int trace_set = 0;
+
+    if (envp)
+    {
+        for (i = 1; i < env_count; ++i)
+        {
+            if ( Strncmp(Krb5_Trace, envp[i], sizeof(Krb5_Trace)-1) == 0 ) 
+            {
+                trace_set = 1;
+                newEnv[new_env_count] = (char*)PAL_Strdup(envp[i]);
+                if (!newEnv[new_env_count])
+                {
+                    return NULL;
+                }
+                new_env_count++;
+            }
+        }
+    }
+
+    // Trace
+    char *strp = NULL;
+
+    // We respect KRB5_TRACE set in the environment, but if it is not set we still set it if the log level
+    // is info or debug
+
+    if (s_opts.logLevel > 2 && !trace_set)
+    {
+        newEnv[new_env_count] = PAL_Malloc(sizeof(Krb5_Trace));
+        if (!newEnv[new_env_count])
+        {
+            return NULL;
+        }
+        strp = newEnv[new_env_count];
+        memcpy(strp, Krb5_Trace, sizeof(Krb5_Trace)-1);
+        new_env_count++;
+    }
+
+    // Keytab
+    newEnv[new_env_count] = PAL_Malloc(sizeof(Krb5_KTName)+strlen(s_opts.krb5KeytabPath));
+    if (!newEnv[new_env_count])
+    {
+        return NULL;
```

Comment:
do we need to free exist mallocated newEnv elements even current element allocate fail?